core/sql/arkcmp/CmpStatement.cpp

```diff
@@ -1688,3 +1690,26 @@ void CmpStatement::setTMUDFRefusedRequirements(const char *details)
 
   detailsOnRefusedRequirements_->insert(new(heap_) NAString(details, heap_));
 }
+
+void CmpStatement::addCSEInfo(CSEInfo *info)
+{
+  if (cses_ == NULL)
+    cses_ = new(CmpCommon::statementHeap())
+      LIST(CSEInfo *)(CmpCommon::statementHeap());
+
+  info->setCSEId(cses_->entries());
+  cses_->insert(info);
+}
+
+CSEInfo * CmpStatement::getCSEInfo(const char *cseName)
+{
+  if (cses_)
+    for (CollIndex i=0; i<cses_->entries(); i++)
+      {
+        if ((*cses_)[i]->getName() == cseName)
```

Comment:
As far as I understand, the WITH clause can appear only at the beginning of a statement and the names must be unique. The scope of these names extends across the entire statement.